### PR TITLE
[DO NOT MERGE] [DO NOT REVIEW] pcm_converter: Replace asserts with sample number check to if instruction

### PR DESCRIPTION
**Pull request by ktrzcinx**
Using if with return instruction is not so fatal consequences
as assertion, so fw will keep running after such an error.

Signed-off-by: Karol Trzcinski <karolx.trzcinski@linux.intel.com>

Fix for https://github.com/thesofproject/sof/issues/3146

### CODE REVIEW

File: src/audio/pcm_converter/pcm_converter.c [hunk 1]
```diff
@@ -28,8 +35,19 @@ void pcm_convert_as_linear(const struct audio_stream *source, uint32_t ioffset,
 	int chunk;
 	int N1, N2;
 
-	assert(audio_stream_get_avail_samples(source) >= samples + ioffset);
-	assert(audio_stream_get_free_samples(sink) >= samples + ooffset);
+	/* assert enough avail/free samples in source and sink buffer */
+	if (audio_stream_get_avail_samples(source) < samples + ioffset) {
+		tr_err(&pcm_conv_tr, "Available samples %d in source < samples to convert %d + offset %d",
+		       audio_stream_get_avail_samples(source), samples,
```
Comment (plbossart) [inline src/audio/pcm_converter/pcm_converter.c:41]: do you really need two calls for this, use a local variable to store the result of the function seems more efficient?

File: src/audio/pcm_converter/pcm_converter.c [hunk 2]
```diff
@@ -28,8 +35,19 @@ void pcm_convert_as_linear(const struct audio_stream *source, uint32_t ioffset,
...
+		       ioffset);
+		return;
+	}
+	if (audio_stream_get_free_samples(sink) < samples + ooffset) {
+		tr_err(&pcm_conv_tr, "Free samples %d in sink < samples to convert %d + offset %d",
+		       audio_stream_get_avail_samples(source), samples,
```
Comment (plbossart) [inline src/audio/pcm_converter/pcm_converter.c:47]: same comment, use a local variable - which would also have avoided the copy-paste mistake and using the wrong function.... Grrr.

File: src/audio/pcm_converter/pcm_converter.c [hunk 3]
```diff
@@ -12,7 +12,14 @@
...
 #include <sof/audio/pcm_converter.h>
-#include <sof/debug/panic.h>
+#include <sof/lib/uuid.h>
+#include <sof/trace/trace.h>
+
+/* bde59d40-0a1d-4205-b85a-9139f2bba316 */
+DECLARE_SOF_RT_UUID("pcm_conv", pcm_conv_uuid, 0xbde59d40, 0x0a1d, 0x4205,
+		    0xb8, 0x5a, 0x91, 0x39, 0xf2, 0xbb, 0xa3, 0x16);
+
+DECLARE_TR_CTX(pcm_conv_tr, SOF_UUID(pcm_conv_uuid), LOG_LEVEL_INFO);
```
Comment (dbaluta) [inline src/audio/pcm_converter/pcm_converter.c:23]: Wait. This change here doesn't look to be related to the commit message. Care to add it in a separate commit?

Comment (plbossart) [reply src/audio/pcm_converter/pcm_converter.c:23]: I had the same initial impression but this seems related - looks required to add the pcm_conv_tr object needed in the trace

Comment (ktrzcinx) [reply src/audio/pcm_converter/pcm_converter.c:23]: Yes, it's needed for `tr_err` function. I added short sentence about it to commit message.

File: src/audio/pcm_converter/pcm_converter.c (hunk 2 above)
Comment (ktrzcinx) [reply src/audio/pcm_converter/pcm_converter.c]: updated

File: src/audio/pcm_converter/pcm_converter.c (hunk 1 above)
Comment (ktrzcinx) [reply src/audio/pcm_converter/pcm_converter.c]: updated

**Comment by paulstelian97** [comment]
> Something fishy is going on, all legacy platforms fail alsa-bat tests and the logger shows tons of errors
> 
> ```
> [4338062316.041667] (             nan) c0 pcm_conv           ......../pcm_converter.c:48   ERROR Free samples 0 in sink < samples to convert 96 + offset 0
> ```
> 
> see https://sof-ci.01.org/sofpr/PR3156/build6552/devicetest/ and e.g. look at the results for Broadwell.

This change shouldn't actually affect what happens, this error should only show up when there was a panic before. Surely wasn't there some recovery code that detected the panic and reloaded the firmware when this happened and the reload made everything happen successfully?

There is indeed a problem, though I feel like this patch only unmasks it rather than causing it.

**Comment by paulstelian97** [comment]
Maybe the biggest issue is that pcm_convert_as_linear returns void and the error is traced _and swallowed_ instead of reported in the upper layers. I'd suggest you take a look into that instead, see if there is a way to _actually_ do the error propagation.

**Comment by lgirdwood** [comment]
@ktrzcinx I've reverted the original PR, can we rework this on top.

File: src/audio/dai.c [hunk 4]
```diff
@@ -98,17 +101,31 @@ static void dai_dma_cb(void *arg, enum notify_id type, void *data)
...
 	} else {
-		dma_buffer_copy_from(dd->dma_buffer, dd->local_buffer,
-				     dd->process, bytes);
+		ret = dma_buffer_copy_from(dd->dma_buffer, dd->local_buffer,
+					   dd->process, bytes);
 
 		buffer_ptr = dd->local_buffer->stream.w_ptr;
 	}
 
+	if (ret < 0) {
```
Comment (lgirdwood) [inline src/audio/dai.c:115]: Can we have an inline comment on what we are checking here.

File: src/audio/host.c [hunk 5]
```diff
@@ -129,13 +129,31 @@ static uint32_t host_dma_get_split(struct host_data *hd, uint32_t bytes)
...
-				     hd->process, bytes);
+		ret = dma_buffer_copy_from(hd->dma_buffer, hd->local_buffer,
+					   hd->process, bytes);
 	else
-		dma_buffer_copy_to(hd->local_buffer, hd->dma_buffer,
-				   hd->process, bytes);
+		ret = dma_buffer_copy_to(hd->local_buffer, hd->dma_buffer,
+					 hd->process, bytes);
+
+	if (ret < 0) {
```
Comment (lgirdwood) [inline src/audio/host.c:144]: likewise here with our action.

File: src/audio/pcm_converter/pcm_converter.c [hunk 6]
```diff
@@ -35,10 +35,12 @@ int pcm_convert_as_linear(const struct audio_stream *source, uint32_t ioffset,
 		return -EINVAL;
 
 	while (i < samples) {
-		/* calculate chunk size */
-		N1 = audio_stream_bytes_without_wrap(source, r_ptr);
-		N2 = audio_stream_bytes_without_wrap(sink, w_ptr);
-		chunk = MIN(N1, N2) * s_size_out;
+		/* calculate chunk size, "">> ffs(x) - 1" is equal "/ x" here */
+		N1 = audio_stream_bytes_without_wrap(source, r_ptr) >>
+		     (ffs(s_size_in) - 1);
```
Comment (lgirdwood) [inline src/audio/pcm_converter/pcm_converter.c:40]: Can you do the (ffs() -1) outside the loop since it looks like is constant.

Comment (ktrzcinx) [reply src/audio/pcm_converter/pcm_converter.c:40]: Updated in https://github.com/thesofproject/sof/pull/3231

File: src/audio/dai.c (hunk 4 above)
Comment (ktrzcinx) [reply src/audio/dai.c:115]: Added in https://github.com/thesofproject/sof/pull/3251

File: src/audio/host.c (hunk 5 above)
Comment (ktrzcinx) [reply src/audio/host.c:144]: Added in https://github.com/thesofproject/sof/pull/3251

**Comment by ktrzcinx** [comment]
SOFCI TEST

**Comment by lgirdwood** [comment]
@ktrzcinx looks like it needs a rebase prior to a retest.

**Review by lgirdwood — COMMENTED** [review]
@ktrzcinx I checked the CI failures and all I could see was
```
449252956.354167] (449252960.000000) c0 hda-dma            ..../intel/hda/hda-dma.c:440  hda-dmac: 4 channel 0 -> copy 0x1060 bytes
[449254143.437500] (     1187.083374) c0 hda-dma            ..../intel/hda/hda-dma.c:440  hda-dmac: 5 channel 0 -> copy 0x180 bytes
[449254183.593750] (       40.156250) c0 hda-dma            ..../intel/hda/hda-dma.c:334  hda-dmac: 7 channel 0 -> copy 0x170 bytes
[449255393.854167] (     1210.260376) c0 hda-dma            ..../intel/hda/hda-dma.c:440  hda-dmac: 5 channel 0 -> copy 0x180 bytes
[449255434.583333] (       40.729168) c0 hda-dma            ..../intel/hda/hda-dma.c:334  hda-dmac: 7 channel 0 -> copy 0x190 bytes
[449256644.010417] (     1209.427124) c0 hda-dma            ..../intel/hda/hda-dma.c:440  hda-dmac: 5 channel 0 -> copy 0x180 bytes
[449256684.531250] (       40.520832) c0 hda-dma            ..../intel/hda/hda-dma.c:334  hda-dmac: 7 channel 0 -> copy 0x180 bytes
[449257893.437500] (     1208.906250) c0 hda-dma            ..../intel/hda/hda-dma.c:440  hda-dmac: 5 channel 0 -> copy 0x180 bytes
[449257934.166667] (       40.729168) c0 hda-dma            ..../intel/hda/hda-dma.c:334  hda-dmac: 7 channel 0 -> copy 0x180 bytes
```
I would have expected the timing to be on 1ms instead of 1.2ms ? Any clues ?

**Comment by paulstelian97** [comment]
@ktrzcinx If this PR is still applicable, I recommend you send it again based on the "main" branch.
